mastermind.py

```diff
@@ -0,0 +1,125 @@
+import random
+import string
+import matplotlib.pyplot as plt
+
+def generateRandomString(length):
+    characters = string.ascii_letters + string.digits + string.punctuation
+    randomString = ''.join(random.choice(characters) for _ in range(length))
+    return randomString
+
+def fitnessFunction(candidate, target):
+    return sum(1 for a, b in zip(candidate, target) if a == b)
+
+def selectParents(population, fitnessScores, numParents):
+    parents = sorted(zip(population, fitnessScores), key=lambda x: x[1], reverse=True)[:numParents]
+    return [p[0] for p in parents]
+
+def crossover(parent1, parent2):
+    crossoverPoint = random.randint(0, len(parent1) - 1)
+    return parent1[:crossoverPoint] + parent2[crossoverPoint:]
+
+def mutate(candidate, mutationRate):
+    characters = string.ascii_letters + string.digits + string.punctuation
+    candidateList = list(candidate)
+    for i in range(len(candidateList)):
+        if random.random() < mutationRate:
+            candidateList[i] = random.choice(characters)
+    return ''.join(candidateList)
+
+def geneticAlgorithm(testString, populationSize, iterations, elitism, crossoverNum, generation, mutationRate):
+    population = [generateRandomString(len(testString)) for _ in range(populationSize)]
+    
+    for i in range(iterations):
+        fitnessScores = [fitnessFunction(individual, testString) for individual in population]
+        nextPopulation = selectParents(population, fitnessScores, elitism)
```

Comment:
Ez egy kicsit eros strategia, hogy csak a legjobb x lehet szulo, de vegulis... szinten az a kerdes, hogy ezt lattad-e valahol, vagy sajat otlet? 
Meg annyi kodszervezesi malor, hogy igazabolha jol latom, a `fitnessscores`-t csak ebben a fuggvenyben hasznalod, szebb lenne, ha a` fitnessfunctiont` adnad at a `selectParents`-nek, es az foglalkozna vele. (Meg ertelmesebb nev lenne a `best_candidates` vagy hasonlo, de persze ez reszletkerdes)

Reply:
Az evolúció azon elvét követtem, hogy a legerősebb szaporodik. Véletlenül mixelni egy esetleg teljesen fals eredménnyel nem annyira célravezető, de tény, hogy nagyobb diverzitást eredményezne.

Reply:
Illetve azért nevezem next_population-nek, mert a következő populációnak ezek lesznek az első egyedei és a next_population-t ezután töltöm fel a (random vagy crossover által) generált egyedekkel. De kivehetem egy best_candidates-nek és azt betehetem a next_population elejére, úgy olvashatóbb.

Reply:
Igazabol nincsenek "torvenyek", ha nem a szokasos, jol mukodik, es hasonlit valamire a termeszetben, mar adhatsz is neki egy uj nevet :-P Viccet felreteve: akkor ha jol ertem, nem lattad mashol, sajat kutfobol jott a dolog.